packages/Kokkos/src/DTK_Mesh_def.hpp

```diff
@@ -0,0 +1,122 @@
+  HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+  LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+  DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+  THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+//---------------------------------------------------------------------------//
+/**
+ * \brief DTK_Mesh_def.hpp
+ */
+//---------------------------------------------------------------------------//
+
+#ifndef DTK_MESH_DEF_HPP
+#define DTK_MESH_DEF_HPP
+
+#include "DTK_ConfigDefs.hpp"
+#include "DTK_DBC.hpp"
+
+namespace DataTransferKit
+{
+template <typename SC, typename LO, typename GO, typename NO>
+Mesh<SC, LO, GO, NO>::Mesh( const Teuchos::RCP<const Teuchos::Comm<int>> &comm,
+                            const global_id_view node_ids,
+                            const connectivity_view connectivity,
+                            const coordinate_view coordinates,
+                            const std::vector<shards::CellTopology> &topology )
+{
+    using connectivity_view_2d =
+        Kokkos::View<local_ordinal_type **, device_type>;
+
+    std::set<shards::CellTopology> topo_done;
+    const size_t n_cells = topology.size();
+    size_t cells_left = topology.size();
+    _mesh_blocks = Teuchos::rcp( new std::vector<MeshBlock<SC, LO, GO, NO>>() );
+    while ( cells_left > 0 )
+    {
+        // Search for a topology that we haven't treated yet
+        size_t topo_pos = 0;
+        while ( topo_done.count( topology[topo_pos] ) == 1 )
+            ++topo_pos;
+        const shards::CellTopology &current_topo = topology[topo_pos];
+        const unsigned int n_nodes = current_topo.getNodeCount();
+
+        // Count the number of cells that have the current topology
+        size_t n_cells_current_topo = 0;
+        for ( size_t i = 0; i < n_cells; ++i )
+            if ( topology[i] == current_topo )
+                ++n_cells_current_topo;
+
+        // Create the MeshBlock for the current_topology
+        size_t pos = 0;
+        size_t cell = 0;
+        // Transform the 1D view of the connectivity to a 2D view
+        connectivity_view_2d connectivity_2d( "connectivity_2d",
+                                              n_cells_current_topo, n_nodes );
+        for ( size_t i = 0; i < topology.size(); ++i )
+        {
+            if ( topology[i].getKey() == current_topo.getKey() )
+            {
+                for ( unsigned int n = 0; n < n_nodes; ++n )
+                    connectivity_2d( cell, n ) = connectivity( pos + n );
+                ++cell;
+                --cells_left;
+                if ( cell == n_cells_current_topo )
+                    break;
+            }
+
+            pos += n_nodes;
```

Comment:
Yeah, that's the plan. I didn't finish the test because I just wanted to show the API and I don't want to spend time writing code that we are not going to use. But that's what I am going to do if we agree  on the interface.